### PR TITLE
Add external output support

### PR DESCRIPTION
**Pull request by JSUYA**
Add feature to support external output in common profile.
The eom API only supports elm_win. So TizenWindowElementary supports this.
Currently, only HDMI type is available.
It can be extended according to the requirements of the Common profile.

https://docs.tizen.org/application/native/guides/device/ext-output/

+)
How to map input device.
1) Check to input event node.
$) winfo -input_devices
2) Set to input and output event node 
$) winfo -input_output_set [input(ex./dev/input/eventX)] [output(ex.HDMIA-1)]

related pr : https://github.com/flutter-tizen/flutter-tizen/pull/465

### CODE REVIEW

File: flutter/shell/platform/tizen/BUILD.gn [hunk 1]
```diff
@@ -160,7 +161,9 @@ template("embedder") {
       libs += [
         "tzsh_common",
         "tzsh_softkey",
+        "eom",
```
Comment (swift-kim) [inline flutter/shell/platform/tizen/BUILD.gn:164]: Please sort alphabetically.

File: flutter/shell/platform/tizen/BUILD.gn [hunk 2]
```diff
@@ -160,7 +161,9 @@ template("embedder") {
...
       ]
+      defines += [ "EOM_SUPPORT" ]
```
Comment (swift-kim) [inline flutter/shell/platform/tizen/BUILD.gn:166]: `defines` appears for the first time in this template at line 169. You may move the whole if block (lines 158-167) to between lines 170 and 172, and also add a blank line between `libs` and `defines`.

Otherwise you can just use the `COMMON_PROFILE` macro without defining `EOM_SUPPORT` itself.

File: flutter/shell/platform/tizen/BUILD.gn (hunk 1 above)
Comment (JSUYA) [reply flutter/shell/platform/tizen/BUILD.gn]: done

File: flutter/shell/platform/tizen/BUILD.gn (hunk 2 above)
Comment (JSUYA) [reply flutter/shell/platform/tizen/BUILD.gn]: Thanks for your comments. COMMON_PROFILE looks better.

File: flutter/shell/platform/tizen/tizen_window_elementary.cc [hunk 3]
```diff
@@ -403,4 +444,74 @@ void TizenWindowElementary::PrepareInputMethod() {
       [this](std::string str) { view_delegate_->OnCommit(str); });
 }
 
+#ifdef EOM_SUPPORT
```
Comment (swift-kim) [inline flutter/shell/platform/tizen/tizen_window_elementary.cc]: Apparently the EOM module itself is available on all profiles. 

https://docs.tizen.org/application/native/api/wearable/4.0/group__CAPI__UI__EOM__MODULE.html
https://docs.tizen.org/application/native/api/mobile/5.5/group__CAPI__UI__EOM__MODULE.html

c.f. This is not the case for `tzsh_softkey` which is not available on wearable 4.0.

How about removing the `#ifdef` guards and making the code common to all profiles?

File: flutter/shell/platform/tizen/tizen_window_elementary.cc [hunk 4]
```diff
@@ -323,8 +364,8 @@ void TizenWindowElementary::UnregisterEventHandlers() {
 }
 
 TizenGeometry TizenWindowElementary::GetGeometry() {
-  // FIXME : evas_object_geometry_get() and ecore_wl2_window_geometry_get() are
-  // not equivalent.
+  // FIXME : evas_object_geometry_get() and ecore_wl2_window_geometry_get()
+  // are not equivalent.
```
Comment (swift-kim) [inline flutter/shell/platform/tizen/tizen_window_elementary.cc:368]: Looks like an unnecessary diff.

File: flutter/shell/platform/tizen/tizen_window_elementary.cc [hunk 5]
```diff
@@ -403,4 +444,74 @@ void TizenWindowElementary::PrepareInputMethod() {
...
+
+int32_t TizenWindowElementary::GetExternalOutputId() {
+  eom_output_id* output_ids = NULL;
+  eom_output_id output_id = 0;
+  eom_output_type_e output_type = EOM_OUTPUT_TYPE_UNKNOWN;
+  int32_t id_cnt = 0;
+
+  output_ids = eom_get_eom_output_ids(&id_cnt);
+  if (id_cnt == 0) {
+    FT_LOG(Error) << "No external outputs supported";
```
Comment (swift-kim) [inline flutter/shell/platform/tizen/tizen_window_elementary.cc:457]: Always end a sentence with a period.

```suggestion
  if (!output_ids || id_cnt == 0) {
    FT_LOG(Error) << "No external output found.";
```

File: flutter/shell/platform/tizen/tizen_window_elementary.cc [hunk 6]
```diff
@@ -403,4 +444,74 @@ void TizenWindowElementary::PrepareInputMethod() {
...
+
+  if (output_ids)
+    free(output_ids);
+
+  return output_id;
+}
+
+bool TizenWindowElementary::InitializeEom() {
+  if (eom_init()) {
+    FT_LOG(Error) << "Eom init fail.";
```
Comment (swift-kim) [inline flutter/shell/platform/tizen/tizen_window_elementary.cc:480]:
```suggestion
    FT_LOG(Error) << "Failed to initialize the External Output Manager.";
```

File: flutter/shell/platform/tizen/tizen_window_elementary.cc [hunk 7]
```diff
@@ -69,11 +93,28 @@ bool TizenWindowElementary::CreateWindow() {
...
-
   int32_t width, height;
-  ecore_evas_screen_geometry_get(ecore_evas, nullptr, nullptr, &width, &height);
+#ifdef EOM_SUPPORT
+  if (external_output_type_ != FlutterDesktopExternalOutputType::kNone) {
+    eom_get_output_resolution(ext_output_id_, &width, &height);
+
+    if (width == 0 || height == 0) {
+      FT_LOG(Error) << "Invalid screen size: " << width << " x " << height;
+      return false;
```
[the remark this answers is not in the document]
Comment (JSUYA) [reply flutter/shell/platform/tizen/tizen_window_elementary.cc]: Removed duplicate code.

File: flutter/shell/platform/tizen/tizen_window_elementary.cc [hunk 8]
```diff
@@ -69,11 +93,28 @@ bool TizenWindowElementary::CreateWindow() {
   elm_win_aux_hint_add(elm_win_, "wm.policy.win.user.geometry", "1");
 #endif
 
-  Ecore_Evas* ecore_evas =
-      ecore_evas_ecore_evas_get(evas_object_evas_get(elm_win_));
-
   int32_t width, height;
```
[the remark this answers is not in the document]
Comment (JSUYA) [reply flutter/shell/platform/tizen/tizen_window_elementary.cc]: done

File: flutter/shell/platform/tizen/tizen_window_elementary.cc [hunk 9]
```diff
@@ -69,11 +93,28 @@ bool TizenWindowElementary::CreateWindow() {
...
+    eom_get_output_resolution(ext_output_id_, &width, &height);
+
+    if (width == 0 || height == 0) {
+      FT_LOG(Error) << "Invalid screen size: " << width << " x " << height;
+      return false;
+    }
+
+    if (eom_set_output_window(ext_output_id_, elm_win_) != EOM_ERROR_NONE) {
+      FT_LOG(Error) << "eom_set_output_window() failed.";
+    }
```
[the remark this answers is not in the document]
Comment (JSUYA) [reply flutter/shell/platform/tizen/tizen_window_elementary.cc]: added `return false`

File: flutter/shell/platform/tizen/tizen_window_elementary.cc (hunk 4 above)
Comment (JSUYA) [reply flutter/shell/platform/tizen/tizen_window_elementary.cc]: done

File: flutter/shell/platform/tizen/tizen_window_elementary.cc [hunk 10]
```diff
@@ -403,4 +444,74 @@ void TizenWindowElementary::PrepareInputMethod() {
       [this](std::string str) { view_delegate_->OnCommit(str); });
 }
 
+#ifdef COMMON_PROFILE
+
+int32_t TizenWindowElementary::GetExternalOutputId() {
+  eom_output_id* output_ids = NULL;
```
[the remark this answers is not in the document]
Comment (JSUYA) [reply flutter/shell/platform/tizen/tizen_window_elementary.cc]: done

File: flutter/shell/platform/tizen/tizen_window_elementary.cc (hunk 5 above)
Comment (JSUYA) [reply flutter/shell/platform/tizen/tizen_window_elementary.cc]: done

File: flutter/shell/platform/tizen/tizen_window_elementary.cc [hunk 11]
```diff
@@ -403,4 +444,74 @@ void TizenWindowElementary::PrepareInputMethod() {
...
+    if (external_output_type_ == FlutterDesktopExternalOutputType::kHDMI &&
+        (output_type == EOM_OUTPUT_TYPE_HDMIA ||
+         output_type == EOM_OUTPUT_TYPE_HDMIB)) {
+      output_id = output_ids[i];
+      break;
+    }
+  }
+
+  if (output_ids)
+    free(output_ids);
```
[the remark this answers is not in the document]
Comment (JSUYA) [reply flutter/shell/platform/tizen/tizen_window_elementary.cc]: done

File: flutter/shell/platform/tizen/tizen_window_elementary.cc (hunk 6 above)
Comment (JSUYA) [reply flutter/shell/platform/tizen/tizen_window_elementary.cc]: I modified the message. And your suggestion message will output from the failure case of `InitializeExternalOutputManager()`.

File: flutter/shell/platform/tizen/tizen_window_elementary.cc [hunk 12]
```diff
@@ -403,4 +444,74 @@ void TizenWindowElementary::PrepareInputMethod() {
...
+  if (eom_init()) {
+    FT_LOG(Error) << "Eom init fail.";
+    return false;
+  }
+
+  int eom_ret;
+  ext_output_id_ = GetExternalOutputId();
+  eom_output_mode_e output_mode = EOM_OUTPUT_MODE_NONE;
+  if (ext_output_id_ == 0) {
+    FT_LOG(Error) << "External output id is NULL.";
```
[the remark this answers is not in the document]
Comment (JSUYA) [reply flutter/shell/platform/tizen/tizen_window_elementary.cc]: done

File: flutter/shell/platform/tizen/tizen_window_elementary.cc [hunk 13]
```diff
@@ -403,4 +444,74 @@ void TizenWindowElementary::PrepareInputMethod() {
...
+    eom_deinit();
+    return false;
+  }
+
+  eom_ret =
```
[the remark this answers is not in the document]
Comment (JSUYA) [reply flutter/shell/platform/tizen/tizen_window_elementary.cc]: done

File: flutter/shell/platform/tizen/tizen_window_elementary.cc [hunk 14]
```diff
@@ -403,4 +444,74 @@ void TizenWindowElementary::PrepareInputMethod() {
...
+      eom_set_output_attribute(ext_output_id_, EOM_OUTPUT_ATTRIBUTE_EXCLUSIVE);
+  if (eom_ret != EOM_ERROR_NONE) {
+    FT_LOG(Error)
+        << "eom_set_output_attribute() failed. Cannot use external output.";
+    eom_deinit();
+    return false;
+  }
+
+  eom_ret = eom_get_output_mode(ext_output_id_, &output_mode);
```
[the remark this answers is not in the document]
Comment (JSUYA) [reply flutter/shell/platform/tizen/tizen_window_elementary.cc]: my mistake. i removed it

File: flutter/shell/platform/tizen/tizen_window_elementary.cc [hunk 15]
```diff
@@ -403,4 +444,74 @@ void TizenWindowElementary::PrepareInputMethod() {
...
+  ext_output_id_ = GetExternalOutputId();
+  eom_output_mode_e output_mode = EOM_OUTPUT_MODE_NONE;
+  if (ext_output_id_ == 0) {
+    FT_LOG(Error) << "External output id is NULL.";
+    eom_deinit();
+    return false;
+  }
+
+  eom_ret =
+      eom_set_output_attribute(ext_output_id_, EOM_OUTPUT_ATTRIBUTE_EXCLUSIVE);
```
[the remark this answers is not in the document]
Comment (JSUYA) [reply flutter/shell/platform/tizen/tizen_window_elementary.cc]: you are right. I modified it to `NORMAL`

File: tools/generate_sysroot.py [hunk 16]
```diff
@@ -109,6 +109,8 @@
   'vulkan-loader-devel',
   'wayland-extension-client-devel',
   'wayland-devel',
+  'libeom',
+  'libeom-devel',
```
[the remark this answers is not in the document]
Comment (JSUYA) [reply tools/generate_sysroot.py]: done

File: flutter/shell/platform/tizen/tizen_window_elementary.h [hunk 17]
```diff
@@ -68,6 +78,12 @@ class TizenWindowElementary : public TizenWindow {
   Evas_Smart_Cb rotation_changed_callback_ = nullptr;
   std::unordered_map<Evas_Callback_Type, Evas_Object_Event_Cb>
       evas_object_callbacks_;
+
+#ifdef COMMON_PROFILE
+  int ext_output_id_;
+  FlutterDesktopExternalOutputType external_output_type_ =
```
[the remark this answers is not in the document]
Comment (JSUYA) [reply flutter/shell/platform/tizen/tizen_window_elementary.h:81]: done

File: flutter/shell/platform/tizen/tizen_window_elementary.cc [hunk 18]
```diff
@@ -403,4 +428,58 @@ void TizenWindowElementary::PrepareInputMethod() {
...
+bool TizenWindowElementary::InitializeExternalOutputManager() {
+  if (eom_init()) {
+    FT_LOG(Error) << "eom_init() failed.";
+    return false;
+  }
+
+  external_output_id_ = GetExternalOutputId();
+  if (external_output_id_ == 0) {
+    FT_LOG(Error) << "Invalid external output ID.";
+    eom_deinit();
```
Comment (swift-kim) [inline flutter/shell/platform/tizen/tizen_window_elementary.cc:466]: Note: The deinit function will be called again by the class destructor.

**Review by swift-kim — COMMENTED** [review]
Overall looks good to me.

**Review by swift-kim — APPROVED** [review]
